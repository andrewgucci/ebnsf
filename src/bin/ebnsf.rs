use ebnsf::parse_ebnf;

use std::path::PathBuf;

fn main() {
    let ebnf = std::fs::read_to_string("test/bnf.ebnf").unwrap();

    let diagram = parse_ebnf(&ebnf).unwrap();

    let output = PathBuf::from("test.svg");

    std::fs::write(&output, diagram.to_string().into_bytes()).unwrap();
}
