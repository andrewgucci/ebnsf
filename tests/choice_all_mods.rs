use ebnsf::parse_ebnf;

mod common;
use common::save_svg;

#[test]
fn choice_all_mods() {
    let ebnf = "\
<star or opt>  ::= <foo>* | <bar>?
<star or star> ::= <foo>* | <bar>*
<star or plus> ::= <foo>* | <bar>+

<plus or opt>  ::= <foo>+ | <bar>?
<plus or star> ::= <foo>+ | <bar>*
<plus or plus> ::= <foo>+ | <bar>+

<opt or opt>  ::= <foo>? | <bar>?
<opt or star> ::= <foo>? | <bar>*
<opt or plus> ::= <foo>? | <bar>+
";

    let svg = parse_ebnf(ebnf).expect("valid grammar").to_string();

    save_svg("choice_all_mods", &svg);

    insta::assert_snapshot!(svg);
}
