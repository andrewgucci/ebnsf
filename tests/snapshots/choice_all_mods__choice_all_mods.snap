---
source: tests/choice_all_mods.rs
expression: svg
---
<svg class="railroad" viewBox="0 0 338 917" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
<style type="text/css">
svg.railroad {
background-color: hsl(30, 20%, 95%);
background-size: 15px 15px;
background-image: linear-gradient(to right, rgba(30, 30, 30, .05) 1px, transparent 1px),
		  linear-gradient(to bottom, rgba(30, 30, 30, .05) 1px, transparent 1px);
}

svg.railroad rect.railroad_canvas {
stroke-width: 0px;
fill: none;
}

svg.railroad path {
stroke-width: 3px;
stroke: black;
fill: none;
}

svg.railroad .debug {
stroke-width: 1px;
stroke: red;
}

svg.railroad text {
font: 14px monospace;
text-anchor: middle;
}

svg.railroad .nonterminal text {
font-weight: bold;
}

svg.railroad text.comment {
font: italic 12px monospace;
}

svg.railroad rect {
stroke-width: 3px;
stroke: black;
fill:hsl(-290, 70%, 90%);
}

svg.railroad g.labeledbox > rect {
stroke-width: 1px;
stroke: grey;
stroke-dasharray: 5px;
fill:rgba(90, 90, 150, .1);
}
</style>
<rect class="railroad_canvas" height="100%" width="100%"/>
<g class="verticalgrid">
<g class="sequence">
<path d=" M 10 34 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<g class="sequence">
<text class="comment" x="78" y="39">
star or opt</text>
<g class="choice">
<path d=" M 132 34 h 24 m 116 0 h 24"/>
<g class="sequence">
<g class="optional">
<path d=" M 156 34 h 24 m -24 0 a 12 12 0 0 0 12 -12 v 0 a 12 12 0 0 1 12 -12 h 68 m -31 0 l -5 -5 m 0 10 l 5 -5 m 31 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 0 12 12 h -24"/>
<g class="repeat">
<path d=" M 180 34 h 12 m 44 0 h 12 m -12 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 1 -12 12 m 0 0 h -44 a 12 12 0 0 1 -12 -12 v 0 a 12 12 0 0 1 12 -12"/>
<g/>
<g class="nonterminal">
<rect height="22" width="44" x="192" y="23"/>
<text x="214" y="39">
foo</text>
</g>
</g>
</g>
</g>
<path d=" M 132 34 a 12 12 0 0 1 12 12 v 22 m 140 0 v -22 a 12 12 0 0 1 12 -12"/>
<path d=" M 144 68 v 12 a 12 12 0 0 0 12 12 m 92 0 h 24 a 12 12 0 0 0 12 -12 v -12"/>
<g class="sequence">
<g class="optional">
<path d=" M 156 92 h 24 m -24 0 a 12 12 0 0 0 12 -12 v 0 a 12 12 0 0 1 12 -12 h 44 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 0 12 12 h -24"/>
<g class="nonterminal">
<rect height="22" width="44" x="180" y="81"/>
<text x="202" y="97">
bar</text>
</g>
</g>
</g>
</g>
<path d=" M 122 34 h 10"/>
</g>
<path d=" M 306 34 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<path d=" M 25 34 h 10"/>
<path d=" M 296 34 h 10"/>
</g>
<g class="sequence">
<path d=" M 10 139 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<g class="sequence">
<text class="comment" x="82" y="144">
star or star</text>
<g class="choice">
<path d=" M 139 139 h 24 m 116 0 h 24"/>
<g class="sequence">
<g class="optional">
<path d=" M 163 139 h 24 m -24 0 a 12 12 0 0 0 12 -12 v 0 a 12 12 0 0 1 12 -12 h 68 m -31 0 l -5 -5 m 0 10 l 5 -5 m 31 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 0 12 12 h -24"/>
<g class="repeat">
<path d=" M 187 139 h 12 m 44 0 h 12 m -12 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 1 -12 12 m 0 0 h -44 a 12 12 0 0 1 -12 -12 v 0 a 12 12 0 0 1 12 -12"/>
<g/>
<g class="nonterminal">
<rect height="22" width="44" x="199" y="128"/>
<text x="221" y="144">
foo</text>
</g>
</g>
</g>
</g>
<path d=" M 139 139 a 12 12 0 0 1 12 12 v 22 m 140 0 v -22 a 12 12 0 0 1 12 -12"/>
<path d=" M 151 173 v 12 a 12 12 0 0 0 12 12 m 116 0 h 0 a 12 12 0 0 0 12 -12 v -12"/>
<g class="sequence">
<g class="optional">
<path d=" M 163 197 h 24 m -24 0 a 12 12 0 0 0 12 -12 v 0 a 12 12 0 0 1 12 -12 h 68 m -31 0 l -5 -5 m 0 10 l 5 -5 m 31 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 0 12 12 h -24"/>
<g class="repeat">
<path d=" M 187 197 h 12 m 44 0 h 12 m -12 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 1 -12 12 m 0 0 h -44 a 12 12 0 0 1 -12 -12 v 0 a 12 12 0 0 1 12 -12"/>
<g/>
<g class="nonterminal">
<rect height="22" width="44" x="199" y="186"/>
<text x="221" y="202">
bar</text>
</g>
</g>
</g>
</g>
</g>
<path d=" M 129 139 h 10"/>
</g>
<path d=" M 313 139 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<path d=" M 25 139 h 10"/>
<path d=" M 303 139 h 10"/>
</g>
<g class="sequence">
<path d=" M 10 257 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<g class="sequence">
<text class="comment" x="82" y="262">
star or plus</text>
<g class="choice">
<path d=" M 139 257 h 24 m 116 0 h 24"/>
<g class="sequence">
<g class="optional">
<path d=" M 163 257 h 24 m -24 0 a 12 12 0 0 0 12 -12 v 0 a 12 12 0 0 1 12 -12 h 68 m -31 0 l -5 -5 m 0 10 l 5 -5 m 31 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 0 12 12 h -24"/>
<g class="repeat">
<path d=" M 187 257 h 12 m 44 0 h 12 m -12 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 1 -12 12 m 0 0 h -44 a 12 12 0 0 1 -12 -12 v 0 a 12 12 0 0 1 12 -12"/>
<g/>
<g class="nonterminal">
<rect height="22" width="44" x="199" y="246"/>
<text x="221" y="262">
foo</text>
</g>
</g>
</g>
</g>
<path d=" M 139 257 a 12 12 0 0 1 12 12 v 22 m 140 0 v -22 a 12 12 0 0 1 12 -12"/>
<path d=" M 151 291 v 0 a 12 12 0 0 0 12 12 m 68 0 h 48 a 12 12 0 0 0 12 -12 v 0"/>
<g class="sequence">
<g class="repeat">
<path d=" M 163 303 h 12 m 44 0 h 12 m -12 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 1 -12 12 m 0 0 h -44 a 12 12 0 0 1 -12 -12 v 0 a 12 12 0 0 1 12 -12"/>
<g/>
<g class="nonterminal">
<rect height="22" width="44" x="175" y="292"/>
<text x="197" y="308">
bar</text>
</g>
</g>
</g>
</g>
<path d=" M 129 257 h 10"/>
</g>
<path d=" M 313 257 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<path d=" M 25 257 h 10"/>
<path d=" M 303 257 h 10"/>
</g>
<g class="sequence">
<path d=" M 10 350 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<g class="sequence">
<text class="comment" x="78" y="355">
plus or opt</text>
<g class="choice">
<path d=" M 132 350 h 24 m 68 0 h 48"/>
<g class="sequence">
<g class="repeat">
<path d=" M 156 350 h 12 m 44 0 h 12 m -12 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 1 -12 12 m 0 0 h -44 a 12 12 0 0 1 -12 -12 v 0 a 12 12 0 0 1 12 -12"/>
<g/>
<g class="nonterminal">
<rect height="22" width="44" x="168" y="339"/>
<text x="190" y="355">
foo</text>
</g>
</g>
</g>
<path d=" M 132 350 a 12 12 0 0 1 12 12 v 22 m 116 0 v -22 a 12 12 0 0 1 12 -12"/>
<path d=" M 144 384 v 12 a 12 12 0 0 0 12 12 m 92 0 h 0 a 12 12 0 0 0 12 -12 v -12"/>
<g class="sequence">
<g class="optional">
<path d=" M 156 408 h 24 m -24 0 a 12 12 0 0 0 12 -12 v 0 a 12 12 0 0 1 12 -12 h 44 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 0 12 12 h -24"/>
<g class="nonterminal">
<rect height="22" width="44" x="180" y="397"/>
<text x="202" y="413">
bar</text>
</g>
</g>
</g>
</g>
<path d=" M 122 350 h 10"/>
</g>
<path d=" M 282 350 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<path d=" M 25 350 h 10"/>
<path d=" M 272 350 h 10"/>
</g>
<g class="sequence">
<path d=" M 10 442 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<g class="sequence">
<text class="comment" x="82" y="447">
plus or star</text>
<g class="choice">
<path d=" M 139 442 h 24 m 68 0 h 72 m -33 0 l -5 -5 m 0 10 l 5 -5 m 33 0"/>
<g class="sequence">
<g class="repeat">
<path d=" M 163 442 h 12 m 44 0 h 12 m -12 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 1 -12 12 m 0 0 h -44 a 12 12 0 0 1 -12 -12 v 0 a 12 12 0 0 1 12 -12"/>
<g/>
<g class="nonterminal">
<rect height="22" width="44" x="175" y="431"/>
<text x="197" y="447">
foo</text>
</g>
</g>
</g>
<path d=" M 139 442 a 12 12 0 0 1 12 12 v 22 m 140 0 v -22 a 12 12 0 0 1 12 -12"/>
<path d=" M 151 476 v 12 a 12 12 0 0 0 12 12 m 116 0 h 0 a 12 12 0 0 0 12 -12 v -12"/>
<g class="sequence">
<g class="optional">
<path d=" M 163 500 h 24 m -24 0 a 12 12 0 0 0 12 -12 v 0 a 12 12 0 0 1 12 -12 h 68 m -31 0 l -5 -5 m 0 10 l 5 -5 m 31 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 0 12 12 h -24"/>
<g class="repeat">
<path d=" M 187 500 h 12 m 44 0 h 12 m -12 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 1 -12 12 m 0 0 h -44 a 12 12 0 0 1 -12 -12 v 0 a 12 12 0 0 1 12 -12"/>
<g/>
<g class="nonterminal">
<rect height="22" width="44" x="199" y="489"/>
<text x="221" y="505">
bar</text>
</g>
</g>
</g>
</g>
</g>
<path d=" M 129 442 h 10"/>
</g>
<path d=" M 313 442 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<path d=" M 25 442 h 10"/>
<path d=" M 303 442 h 10"/>
</g>
<g class="sequence">
<path d=" M 10 547 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<g class="sequence">
<text class="comment" x="82" y="552">
plus or plus</text>
<g class="choice">
<path d=" M 139 547 h 24 m 68 0 h 24"/>
<g class="sequence">
<g class="repeat">
<path d=" M 163 547 h 12 m 44 0 h 12 m -12 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 1 -12 12 m 0 0 h -44 a 12 12 0 0 1 -12 -12 v 0 a 12 12 0 0 1 12 -12"/>
<g/>
<g class="nonterminal">
<rect height="22" width="44" x="175" y="536"/>
<text x="197" y="552">
foo</text>
</g>
</g>
</g>
<path d=" M 139 547 a 12 12 0 0 1 12 12 v 22 m 92 0 v -22 a 12 12 0 0 1 12 -12"/>
<path d=" M 151 581 v 0 a 12 12 0 0 0 12 12 m 68 0 h 0 a 12 12 0 0 0 12 -12 v 0"/>
<g class="sequence">
<g class="repeat">
<path d=" M 163 593 h 12 m 44 0 h 12 m -12 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 1 -12 12 m 0 0 h -44 a 12 12 0 0 1 -12 -12 v 0 a 12 12 0 0 1 12 -12"/>
<g/>
<g class="nonterminal">
<rect height="22" width="44" x="175" y="582"/>
<text x="197" y="598">
bar</text>
</g>
</g>
</g>
</g>
<path d=" M 129 547 h 10"/>
</g>
<path d=" M 265 547 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<path d=" M 25 547 h 10"/>
<path d=" M 255 547 h 10"/>
</g>
<g class="sequence">
<path d=" M 10 653 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<g class="sequence">
<text class="comment" x="75" y="658">
opt or opt</text>
<g class="choice">
<path d=" M 125 653 h 24 m 92 0 h 24"/>
<g class="sequence">
<g class="optional">
<path d=" M 149 653 h 24 m -24 0 a 12 12 0 0 0 12 -12 v 0 a 12 12 0 0 1 12 -12 h 44 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 0 12 12 h -24"/>
<g class="nonterminal">
<rect height="22" width="44" x="173" y="642"/>
<text x="195" y="658">
foo</text>
</g>
</g>
</g>
<path d=" M 125 653 a 12 12 0 0 1 12 12 v 9 m 116 0 v -9 a 12 12 0 0 1 12 -12"/>
<path d=" M 137 674 v 12 a 12 12 0 0 0 12 12 m 92 0 h 0 a 12 12 0 0 0 12 -12 v -12"/>
<g class="sequence">
<g class="optional">
<path d=" M 149 698 h 24 m -24 0 a 12 12 0 0 0 12 -12 v 0 a 12 12 0 0 1 12 -12 h 44 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 0 12 12 h -24"/>
<g class="nonterminal">
<rect height="22" width="44" x="173" y="687"/>
<text x="195" y="703">
bar</text>
</g>
</g>
</g>
</g>
<path d=" M 115 653 h 10"/>
</g>
<path d=" M 275 653 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<path d=" M 25 653 h 10"/>
<path d=" M 265 653 h 10"/>
</g>
<g class="sequence">
<path d=" M 10 745 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<g class="sequence">
<text class="comment" x="78" y="750">
opt or star</text>
<g class="choice">
<path d=" M 132 745 h 24 m 92 0 h 48"/>
<g class="sequence">
<g class="optional">
<path d=" M 156 745 h 24 m -24 0 a 12 12 0 0 0 12 -12 v 0 a 12 12 0 0 1 12 -12 h 44 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 0 12 12 h -24"/>
<g class="nonterminal">
<rect height="22" width="44" x="180" y="734"/>
<text x="202" y="750">
foo</text>
</g>
</g>
</g>
<path d=" M 132 745 a 12 12 0 0 1 12 12 v 9 m 140 0 v -9 a 12 12 0 0 1 12 -12"/>
<path d=" M 144 766 v 12 a 12 12 0 0 0 12 12 m 116 0 h 0 a 12 12 0 0 0 12 -12 v -12"/>
<g class="sequence">
<g class="optional">
<path d=" M 156 790 h 24 m -24 0 a 12 12 0 0 0 12 -12 v 0 a 12 12 0 0 1 12 -12 h 68 m -31 0 l -5 -5 m 0 10 l 5 -5 m 31 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 0 12 12 h -24"/>
<g class="repeat">
<path d=" M 180 790 h 12 m 44 0 h 12 m -12 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 1 -12 12 m 0 0 h -44 a 12 12 0 0 1 -12 -12 v 0 a 12 12 0 0 1 12 -12"/>
<g/>
<g class="nonterminal">
<rect height="22" width="44" x="192" y="779"/>
<text x="214" y="795">
bar</text>
</g>
</g>
</g>
</g>
</g>
<path d=" M 122 745 h 10"/>
</g>
<path d=" M 306 745 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<path d=" M 25 745 h 10"/>
<path d=" M 296 745 h 10"/>
</g>
<g class="sequence">
<path d=" M 10 850 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<g class="sequence">
<text class="comment" x="78" y="855">
opt or plus</text>
<g class="choice">
<path d=" M 132 850 h 24 m 92 0 h 24"/>
<g class="sequence">
<g class="optional">
<path d=" M 156 850 h 24 m -24 0 a 12 12 0 0 0 12 -12 v 0 a 12 12 0 0 1 12 -12 h 44 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 0 12 12 h -24"/>
<g class="nonterminal">
<rect height="22" width="44" x="180" y="839"/>
<text x="202" y="855">
foo</text>
</g>
</g>
</g>
<path d=" M 132 850 a 12 12 0 0 1 12 12 v 9 m 116 0 v -9 a 12 12 0 0 1 12 -12"/>
<path d=" M 144 871 v 0 a 12 12 0 0 0 12 12 m 68 0 h 24 a 12 12 0 0 0 12 -12 v 0"/>
<g class="sequence">
<g class="repeat">
<path d=" M 156 883 h 12 m 44 0 h 12 m -12 0 a 12 12 0 0 1 12 12 v 0 a 12 12 0 0 1 -12 12 m 0 0 h -44 a 12 12 0 0 1 -12 -12 v 0 a 12 12 0 0 1 12 -12"/>
<g/>
<g class="nonterminal">
<rect height="22" width="44" x="168" y="872"/>
<text x="190" y="888">
bar</text>
</g>
</g>
</g>
</g>
<path d=" M 122 850 h 10"/>
</g>
<path d=" M 282 850 a 5 5 0 0 1 5 -5 a 5 5 0 0 1 5 5 a 5 5 0 0 1 -5 5 a 5 5 0 0 1 -5 -5 m 10 0 h 5"/>
<path d=" M 25 850 h 10"/>
<path d=" M 272 850 h 10"/>
</g>
</g>
</svg>
